framework/src/utils/MooseCoordTransform.C

```diff
@@ -0,0 +1,276 @@
+    else
+      mooseAssert(false, "we should never get here");
+#endif
+  }
+#ifndef NDEBUG
+  else
+    mooseAssert(false, "we should never get here");
+#endif
+
+  setRotation(alpha, beta, gamma);
+
+  if (bad_rz_rotation)
+    mooseError("Rotation yields negative radial values");
+}
+
+void
+MooseCoordTransform::setRotation(const Real alpha, const Real beta, const Real gamma)
+{
+  _rotate = std::make_unique<RealTensorValue>(
+      RealTensorValue::extrinsic_rotation_matrix(alpha, beta, gamma));
+}
+
+void
+MooseCoordTransform::setRotation(const Real alpha,
+                                 const Real beta,
+                                 const Real gamma,
+                                 const Direction rz_z_axis)
+{
+  if (rz_z_axis == Z)
+    mooseError("Invalid rz_coord_axis option of 'Z'");
+  else if (rz_z_axis == INVALID)
+    mooseError("Called MooseCoordTransform::setRotation with INVALID 'rz_z_axis'");
+
+  const Direction rz_r_axis = rz_z_axis == X ? Y : X;
+
+  bool supported_manual_rz_rotation = false;
+  const auto angles = std::make_tuple(alpha, beta, gamma);
+  if (angles == std::make_tuple(0, 90, 0))
+  {
+    if (rz_r_axis == X)
+    {
+      _rz_r_axis = X;
+      _rz_z_axis = Z;
+    }
+    else if (rz_r_axis == Y)
+    {
+      _rz_r_axis = Z;
+      _rz_z_axis = X;
+    }
+    supported_manual_rz_rotation = true;
+  }
+
+  _rotate = std::make_unique<RealTensorValue>(
+      RealTensorValue::extrinsic_rotation_matrix(alpha, beta, gamma));
+
+  if (!supported_manual_rz_rotation)
+    mooseError("Unsupported manual angle prescription in 'MooseCoordTransform::setRotation'");
+}
+
+void
+MooseCoordTransform::setLengthUnitsPerMeter(const Real length_units_per_meter)
+{
+  const auto scale = 1 / length_units_per_meter;
+
+  _scale =
+      std::make_unique<RealTensorValue>(RealTensorValue(scale, 0, 0, 0, scale, 0, 0, 0, scale));
+}
+
+void
+MooseCoordTransform::setCoordinateSystem(const Moose::CoordinateSystemType coord_type)
+{
+  _coord_type = coord_type;
+}
+
+MooseCoordTransform::MooseCoordTransform(const InputParameters & params)
+{
+  //
+  // coordinate system transformation. If we have multiple different coordinate system types in our
+  // problem, we take note of it because that can cause issues if there is a non-Cartesian
+  // destination coordinate system
+  //
+  const auto & mesh = *params.getCheckedPointerParam<MooseMesh *>("mesh");
+  const auto & coord_sys = mesh.getCoordSystem();
+  std::unordered_set<Moose::CoordinateSystemType> coord_types;
+  auto map_it = coord_sys.begin();
+  // It's possible that the mesh is not in a complete state
+  if (map_it == coord_sys.end())
+    setCoordinateSystem(Moose::COORD_XYZ);
+  else
+    setCoordinateSystem(map_it->second);
+  for (; map_it != coord_sys.end(); ++map_it)
+    coord_types.insert(map_it->second);
+
+  _has_different_coord_sys = coord_types.size() > 1;
+
+  //
+  // rotation
+  //
+  const bool has_alpha = params.isParamValid("alpha_rotation");
```

Comment:
Maybe put the parameters this guy expects in `MooseCordTransform`?

Something like

```
static InputParameters transformParameters();
```

?